sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala

```diff
@@ -481,10 +610,10 @@ class TungstenAggregationIterator(
       // When needsProcess is false, the format of input rows is groupingKey + aggregation buffer.
       // We need to project the aggregation buffer part from an input row.
       val buffer = createNewAggregationBuffer()
-      // The originalInputAttributes are using cloneBufferAttributes. So, we need to use
-      // allAggregateFunctions.flatMap(_.cloneBufferAttributes).
+      // The originalInputAttributes are using inputAggBufferAttributes. So, we need to use
+      // allAggregateFunctions.flatMap(_.inputAggBufferAttributes).
       val bufferExtractor = newMutableProjection(
-        allAggregateFunctions.flatMap(_.inputAggBufferAttributes),
+        originalInputAttributes.drop(initialInputBufferOffset),
```

Comment:
Had a discussion with @JoshRosen offline. Here is the explaining for this change.

Our `ImperativeAggregate` and `DeclarativeAggregate` are quite different. For an `ImperativeAggregate`, its correctness is based on the correctness of `mutableAggBufferOffset` and `inputAggBufferOffset`, and when there is no fallback, it does not rely on attribute ids for binding. To make `mutableAggBufferOffset` and `inputAggBufferOffset` immutable, whenever we call `withNewMutableAggBufferOffset` and `withNewInputAggBufferOffset`, we create a new copy of the `ImperativeAggregate`. Although a new copy contains a new set of attribute ids for `aggBufferAttributes` and `inputAggBufferAttributes`, in most cases, those new ids do not introduce any problem because an `ImperativeAggregate` does not rely on these ids. However, once we fallback to sort-based aggregation and we need to extract input aggregation buffers, we need to have matching attribute ids to make it work. Since  `allAggregateFunctions.flatMap(_.inputAggBufferAttributes)` will have attributes with a new set of ids, we will see a binding error at here. `originalInputAttributes.drop(initialInputBufferOffset)` is fine at here.